dpnp/backend/extensions/lapack/gesv.cpp

```diff
@@ -273,6 +273,162 @@ std::pair<sycl::event, sycl::event>
+        throw py::value_error(
+            "The arrays of coefficients and dependent variables "
+            "are overlapping segments of memory.");
+    }
+
+    bool is_coeff_matrix_f_contig = coeff_matrix.is_f_contiguous();
+    if (!is_coeff_matrix_f_contig) {
+        throw py::value_error("The coefficient matrix "
+                              "must be F-contiguous.");
+    }
+
+    bool is_dependent_vals_f_contig = dependent_vals.is_f_contiguous();
+    if (!is_dependent_vals_f_contig) {
+        throw py::value_error("The array of dependent variables "
+                              "must be F-contiguous.");
+    }
+
+    auto array_types = dpctl_td_ns::usm_ndarray_types();
+    int coeff_matrix_type_id =
+        array_types.typenum_to_lookup_id(coeff_matrix.get_typenum());
+    int dependent_vals_type_id =
+        array_types.typenum_to_lookup_id(dependent_vals.get_typenum());
+
+    if (coeff_matrix_type_id != dependent_vals_type_id) {
+        throw py::value_error("The types of the coefficient matrix and "
+                              "dependent variables are mismatched.");
+    }
+
+    gesv_impl_fn_ptr_t gesv_fn = gesv_dispatch_vector[coeff_matrix_type_id];
+    if (gesv_fn == nullptr) {
+        throw py::value_error(
+            "No gesv implementation defined for the provided type "
+            "of the coefficient matrix.");
+    }
+
+    char *coeff_matrix_data = coeff_matrix.get_data();
+    char *dependent_vals_data = dependent_vals.get_data();
+
+    const std::int64_t batch_size = coeff_matrix_shape[2];
+    const std::int64_t n = coeff_matrix_shape[1];
+    const std::int64_t nrhs =
+        (dependent_vals_nd > 2) ? dependent_vals_shape[1] : 1;
+
+    const std::int64_t lda = std::max<size_t>(1UL, n);
+    const std::int64_t ldb = std::max<size_t>(1UL, n);
+
+    int coeff_matrix_elemsize = coeff_matrix.get_elemsize();
+    int dependent_vals_elemsize = dependent_vals.get_elemsize();
+
+    std::vector<sycl::event> host_task_events;
+    std::vector<sycl::event> gesv_task_events;
+
+    host_task_events.reserve(batch_size);
+    gesv_task_events.reserve(batch_size);
+
+    {
+        // Release GIL to allow other Python threads to run during the loop
+        // as the operations in the loop do not require GIL
+        py::gil_scoped_release release;
+
+        for (std::int64_t i = 0; i < batch_size; ++i) {
+            char *coeff_matrix_batch =
+                coeff_matrix_data + i * n * n * coeff_matrix_elemsize;
+            char *dependent_vals_batch =
+                dependent_vals_data + i * n * nrhs * dependent_vals_elemsize;
+
+            sycl::event gesv_ev =
+                gesv_fn(exec_q, n, nrhs, coeff_matrix_batch, lda,
+                        dependent_vals_batch, ldb, host_task_events, depends);
+
+            gesv_task_events.push_back(gesv_ev);
+        }
+    }
+
+    sycl::event combine_ev = exec_q.submit(
+        [&](sycl::handler &cgh) { cgh.depends_on(gesv_task_events); });
```

Comment:
What is the reason to combine them? Would not it be better to return a vector of events, like:
```cpp
return std::make_pair(args_ev, gesv_task_events);
```
I guess extra kernel submission might impact the performance.

Reply:
The implementation has been changed.
Now `gesv` and `gesv_batch` return executing event and host event